src/ts_catalog/continuous_aggs_watermark.c

```diff
@@ -0,0 +1,319 @@
+
+	if (watermark_cache != NULL)
+	{
+		if (cagg_watermark_is_valid(watermark_cache, mat_hypertable_id))
+			PG_RETURN_INT64(watermark_cache->value);
+
+		MemoryContextDelete(watermark_cache->mctx);
+	}
+
+	cagg = ts_continuous_agg_find_by_mat_hypertable_id(mat_hypertable_id);
+
+	if (NULL == cagg)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid materialized hypertable ID: %d", mat_hypertable_id)));
+
+	/* Preemptive permission check to ensure the function complains about lack
+	 * of permissions on the cagg rather than the materialized hypertable */
+	aclresult = pg_class_aclcheck(cagg->relid, GetUserId(), ACL_SELECT);
+	aclcheck_error(aclresult, OBJECT_MATVIEW, get_rel_name(cagg->relid));
+	watermark_cache = cagg_watermark_create(cagg, TopTransactionContext);
+
+	PG_RETURN_INT64(watermark_cache->value);
+}
+
+static ScanTupleResult
+cagg_watermark_scan_update(TupleInfo *ti, void *data)
+{
+	int64 *new_watermark = data;
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	HeapTuple new_tuple = heap_copytuple(tuple);
+	Form_continuous_aggs_watermark form = (Form_continuous_aggs_watermark) GETSTRUCT(new_tuple);
+
+	form->watermark = *new_watermark;
+	ts_catalog_update(ti->scanrel, new_tuple);
+	heap_freetuple(new_tuple);
+
+	if (should_free)
+		heap_freetuple(tuple);
+
+	return SCAN_DONE;
+}
+
+static void
+cagg_watermark_set_internal(int32 mat_hypertable_id, int64 new_watermark)
+{
+	bool watermark_updated;
+	ScanKeyData scankey[1];
+
+	ScanKeyInit(&scankey[0],
+				Anum_continuous_aggs_watermark_mat_hypertable_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(mat_hypertable_id));
+
+	watermark_updated = ts_catalog_scan_one(CONTINUOUS_AGGS_WATERMARK /*=table*/,
+											CONTINUOUS_AGGS_WATERMARK_PKEY /*=indexid*/,
+											scankey /*=scankey*/,
+											1 /*=num_keys*/,
+											cagg_watermark_scan_update /*=tuple_found*/,
+											RowExclusiveLock /*=lockmode*/,
+											CONTINUOUS_AGGS_WATERMARK_TABLE_NAME /*=table_name*/,
+											&new_watermark /*=data*/);
+
+	if (!watermark_updated)
+	{
+		Catalog *catalog = ts_catalog_get();
+		Relation rel =
+			table_open(catalog_get_table_id(catalog, CONTINUOUS_AGGS_WATERMARK), RowExclusiveLock);
+		TupleDesc desc = RelationGetDescr(rel);
+		Datum values[Natts_continuous_aggs_watermark];
+		bool nulls[Natts_continuous_aggs_watermark] = { false };
+		CatalogSecurityContext sec_ctx;
+
+		values[AttrNumberGetAttrOffset(Anum_continuous_aggs_watermark_mat_hypertable_id)] =
+			Int32GetDatum(mat_hypertable_id);
+		values[AttrNumberGetAttrOffset(Anum_continuous_aggs_watermark_watermark)] =
+			Int64GetDatum(new_watermark);
+
+		ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+		ts_catalog_insert_values(rel, desc, values, nulls);
+		ts_catalog_restore_user(&sec_ctx);
+
+		table_close(rel, RowExclusiveLock);
```

Comment:
Since you're releasing the `RowExclusiveLock` here another transaction that started earlier might come in and change the value to a watermark that is not the correct one. I think you need to keep the lock until the end of the transaction.